Cellular: Gemalto Cinterion support for ELS61 and BGS2

AriParkkila opened the pull request:
### Description

AT command support for the Gemalto Cinterion cellular modules:
* LWIP supported in PPP mode
* UDP sockets supported in AT mode

Depends on PR #7860.

### Pull request type

    [ ] Fix
    [ ] Refactor
    [ ] New target
    [X] Feature
    [ ] Breaking change

Commit:
Cellular: Gemalto Cinterion support for ELS61

AriParkkila commented:
@jarvte @mirelachirica @TeemuKultala please review

jarvte commented on features/cellular/framework/targets/GEMALTO/CINTERION/GEMALTO_CINTERION_CellularNetwork.cpp:
```diff
@@ -0,0 +1,58 @@
+    return (reg_type == C_REG || reg_type == C_GREG || reg_type == C_EREG);
+}
+
+const char *GEMALTO_CINTERION_CellularNetwork::get_apn() const
```
no need for this method as _apn is protected in inherited class

jarvte commented on features/cellular/framework/targets/GEMALTO/CINTERION/GEMALTO_CINTERION_CellularStack.cpp:
```diff
@@ -0,0 +1,491 @@
+    }
+
+    socket->created = true;
+    tr_info("Socket %d created (err %d)", socket->id, _at.get_last_error());
```
Could be debug so that log is not polluted.

jarvte commented on features/cellular/framework/targets/GEMALTO/CINTERION/GEMALTO_CINTERION_CellularStack.cpp:
```diff
@@ -0,0 +1,491 @@
+nsapi_size_or_error_t GEMALTO_CINTERION_CellularStack::socket_sendto_impl(CellularSocket *socket,
+                                                                          const SocketAddress &address, const void *data, nsapi_size_t size)
+{
+    tr_info("Socket %d, sendto %s, len %d", socket->id, address.get_ip_address(), size);
```
Could be debug so that log is not polluted.

jarvte commented on features/cellular/framework/targets/GEMALTO/CINTERION/GEMALTO_CINTERION_CellularStack.cpp:
```diff
@@ -0,0 +1,491 @@
+    _at.write_bytes((uint8_t *)data, accept_len);
+    _at.resp_stop();
+
+    tr_info("Socket %d sendto %s, %d bytes (err %d)", socket->id, address.get_ip_address(), accept_len, _at.get_last_error());
```
Could be debug so that log is not polluted.

jarvte commented on features/cellular/framework/targets/GEMALTO/CINTERION/GEMALTO_CINTERION_CellularStack.cpp:
```diff
@@ -0,0 +1,491 @@
+nsapi_size_or_error_t GEMALTO_CINTERION_CellularStack::socket_recvfrom_impl(CellularSocket *socket, SocketAddress *address,
+                                                                            void *buffer, nsapi_size_t size)
+{
+    tr_info("Socket %d recvfrom %d bytes", socket->id, size);
```
Could be debug so that log is not polluted.

jarvte commented on features/cellular/framework/targets/GEMALTO/CINTERION/GEMALTO_CINTERION_CellularStack.cpp:
```diff
@@ -0,0 +1,491 @@
+
+    _at.resp_stop();
+
+    tr_info("Socket %d, recvfrom %s, %d bytes (err %d)", socket->id, ip_address, len, _at.get_last_error());
```
Could be debug so that log is not polluted.

AriParkkila commented:
Removed redundant `get_apn()` and fixed tracing level to `tr_debug`. Also rebased with mbed-os/master.

Commit:
Cellular: Added support for the Gemalto/BGS2 cellular module

Commit:
Cellular: Added support for the Gemalto/EMS31 cellular module

cmonr commented:
/morph build

0xc0170 commented:
Restarting the build, should be fine now (previous failure was known yesterday and was fixed)

/morph build

cmonr commented:
Pausing Test CI until 5.9.6 PR is merged.
Will restart CI shortly after.

cmonr commented:
/morph test

cmonr commented:
/morph uvisor-test

0xc0170 commented:
/morph uvisor-test

0xc0170 commented:
uvisor CI reports this error : `GEMALTO_CINTERION_CellularStack.cpp@461,28: 'strtol' is not a member of 'std'` please review

AriParkkila commented:
@0xc0170 fixed missing std::strtol by including <cstdlib>

0xc0170 commented:
/morph build

cmonr commented:
/morph test

0xc0170 commented on features/cellular/framework/targets/GEMALTO/CINTERION/GEMALTO_CINTERION_CellularStack.cpp:
```diff
@@ -0,0 +1,558 @@
+{
+    int sock_id = _at.read_int();
+    int urc_code = _at.read_int();
+    //tr_debug("RX event: socket=%d, urc=%d", sock_id, urc_code);
```
dead code can be removed

0xc0170 commented on features/cellular/framework/targets/GEMALTO/CINTERION/GEMALTO_CINTERION_CellularStack.cpp:
```diff
@@ -0,0 +1,558 @@
+            char paramTag[16];
+            int paramTagLen = _at.read_string(paramTag, sizeof(paramTag));
+            if (paramTagLen > 0) {
+                //tr_debug("paramTag %s", paramTag);
```
Probably few more lines left here that could be removed

0xc0170 commented on features/cellular/framework/targets/GEMALTO/CINTERION/GEMALTO_CINTERION_CellularStack.h:
```diff
@@ -0,0 +1,61 @@
+                                                       void *buffer, nsapi_size_t size);
+
+private:
+    CellularSocket *find_socket(int sock_id);
```
shall this private methods be documented - at least one liners (developers might be interested).

0xc0170 commented:
I had the last look at the changes. The main one is lines that are commented out (debug info). Please let us know. They shall be fixed. (Can be subsequent PR in this case, as it's cosmetic change request)

Commit:
Cellular: Fixed network registration on Gemalto AT drivers

AriParkkila commented:
@0xc0170 added more comments and removed out-commented code. Also fixed network registration as that has changed in PR #7860, which must be merged first.

cmonr commented:
@AriParkkila Were a couple of lines missed?

Normally GitHub will collapse review comments when the line disappears from the review.

AriParkkila commented:
@cmonr for me, GitHub is a bit too clever in how it handles review comments, do you see something missing there... I'm guessing that due to this PR does not compile without PR #7860 that might mess up how GitHub shows review comments?

AriParkkila commented:
Please trigger morph builds to see that this passes all tests.

cmonr commented:
/morph build
